comp/networkpath/npscheduler/npschedulerimpl/npscheduler.go

```diff
@@ -7,31 +7,324 @@
-	return &npSchedulerImpl{
-		enabled: false,
+func shouldScheduleNetworkPathForConn(conn *model.Connection) bool {
+	return conn.Direction == model.ConnectionDirection_outgoing
+}
+
+// scheduleOne schedules pathtests.
+// It shouldn't block, if the input channel is full, an error is returned.
+func (s *npSchedulerImpl) scheduleOne(hostname string, port uint16) error {
+	if s.pathtestInputChan == nil {
+		return errors.New("no input channel, please check that network path is enabled")
+	}
+	s.logger.Debugf("Schedule traceroute for: hostname=%s port=%d", hostname, port)
+
+	if net.ParseIP(hostname).To4() == nil {
+		// TODO: IPv6 not supported yet
+		s.logger.Debugf("Only IPv4 is currently supported. Address not supported: %s", hostname)
+		return nil
 	}
+
+	ptest := &common.Pathtest{
+		Hostname: hostname,
+		Port:     port,
+	}
+	select {
+	case s.pathtestInputChan <- ptest:
+		return nil
+	default:
+		return fmt.Errorf("scheduler input channel is full (channel capacity is %d)", cap(s.pathtestInputChan))
+	}
+}
+func (s *npSchedulerImpl) start() error {
+	if s.running {
+		return errors.New("server already started")
+	}
+	s.running = true
+	s.logger.Info("Start NpScheduler")
+	go s.listenPathtests()
+	go s.flushLoop()
+	s.startWorkers()
+	return nil
 }
 
-func newNpSchedulerImpl(epForwarder eventplatform.Component) *npSchedulerImpl {
-	return &npSchedulerImpl{
-		enabled:     true,
-		epForwarder: epForwarder,
+func (s *npSchedulerImpl) stop() {
+	s.logger.Info("Stop NpScheduler")
+	if !s.running {
+		return
+	}
+	close(s.stopChan)
+	<-s.flushLoopDone
+	<-s.runDone
+	s.running = false
+}
+
+func (s *npSchedulerImpl) listenPathtests() {
+	s.logger.Debug("Starting listening for pathtests")
+	for {
+		select {
+		case <-s.stopChan:
+			s.logger.Info("Stopped listening for pathtests")
+			s.runDone <- struct{}{}
+			return
+		case ptest := <-s.pathtestInputChan:
+			s.logger.Debugf("Pathtest received: %+v", ptest)
+			s.receivedPathtestCount.Inc()
+			s.pathtestStore.Add(ptest)
+		}
+	}
+}
+
+func (s *npSchedulerImpl) runTracerouteForPath(ptest *pathteststore.PathtestContext) {
+	s.logger.Debugf("Run Traceroute for ptest: %+v", ptest)
+
+	startTime := time.Now()
+	cfg := traceroute.Config{
+		DestHostname: ptest.Pathtest.Hostname,
+		DestPort:     ptest.Pathtest.Port,
+		MaxTTL:       0, // TODO: make it configurable, setting 0 to use default value for now
+		TimeoutMs:    0, // TODO: make it configurable, setting 0 to use default value for now
+	}
+
+	path, err := s.runTraceroute(cfg)
+	if err != nil {
+		s.logger.Errorf("%s", err)
+		return
+	}
+
+	s.sendTelemetry(path, startTime, ptest)
+
+	payloadBytes, err := json.Marshal(path)
+	if err != nil {
+		s.logger.Errorf("json marshall error: %s", err)
+	} else {
+		s.logger.Debugf("network path event: %s", string(payloadBytes))
+		m := message.NewMessage(payloadBytes, nil, "", 0)
+		err = s.epForwarder.SendEventPlatformEventBlocking(m, eventplatform.EventTypeNetworkPath)
```

Comment:
Do we want to make this asynchronous?

Reply:
The name `SendEventPlatformEventBlocking` might not be very clear.

"blocking" here doesn't mean synchronous. The different between `SendEventPlatformEvent` and `SendEventPlatformEventBlocking` are:

```
// SendEventPlatformEvent sends messages to the event platform intake.
// SendEventPlatformEvent will drop messages and return an error if the input channel is already full.
```

```
// SendEventPlatformEventBlocking sends messages to the event platform intake.
// SendEventPlatformEventBlocking will block if the input channel is already full.
```

Code details of the differences between those 2 functions are here: https://github.com/DataDog/datadog-agent/blob/00a98b2da38e2190143698986e6ed578087813a0/comp/forwarder/eventplatform/eventplatformimpl/epforwarder.go

In our case here, I think we prefer avoid dropping if possible, and blocking when the epForwarder input channel is full should be fine. The only consequence is to make instances of the NpScheduler workers slower.